src/lib/xmlparse.rs

```diff
@@ -1101,38 +1111,272 @@ macro_rules! hash_lookup {
 }
 
 #[repr(C)]
-#[derive(Clone)]
-pub struct DTD<'scf> {
+pub struct DTD<'p> {
+    tables: RefCell<DTDTables>,
+    pools: RefCell<DTDPools>,
+    scaffold: Cow<'p, RefCell<DTDScaffold>>,
+    keepProcessing: Cell<XML_Bool>,
+    hasParamEntityRefs: Cell<XML_Bool>,
+    standalone: Cell<XML_Bool>,
+    paramEntityRead: Cell<XML_Bool>,
+    defaultPrefix: Cell<PREFIX>,
+    in_eldecl: Cell<XML_Bool>,
+    contentStringLen: Cell<c_uint>,
+}
+
+#[derive(Default)]
+pub struct DTDTables {
     // TODO: get rid of the `Box`es to eliminate the extra indirection;
     // for now, we can keep them since they're equivalent to the C code's
     // structure anyway
-    pub generalEntities: HashMap<HashKey, Box<ENTITY>>,
-    pub elementTypes: HashMap<HashKey, Box<ELEMENT_TYPE>>,
-    pub attributeIds: HashMap<HashKey, Box<ATTRIBUTE_ID>>,
-    pub prefixes: HashMap<HashKey, Box<PREFIX>>,
-    pub pool: STRING_POOL,
-    pub entityValuePool: STRING_POOL,
-    pub keepProcessing: XML_Bool,
-    pub hasParamEntityRefs: XML_Bool,
-    pub standalone: XML_Bool,
-    pub paramEntityRead: XML_Bool,
+    generalEntities: HashMap<HashKey, Box<ENTITY>>,
+    elementTypes: HashMap<HashKey, Box<ELEMENT_TYPE>>,
+    attributeIds: HashMap<HashKey, Box<ATTRIBUTE_ID>>,
+    prefixes: HashMap<HashKey, Box<PREFIX>>,
     // `test_alloc_nested_entities` counts the allocations,
     // so we need to use `Box` here to pass that test
-    pub paramEntities: HashMap<HashKey, Box<ENTITY>>,
-    pub defaultPrefix: PREFIX,
-    pub in_eldecl: XML_Bool,
-    pub scaffold: Cow<'scf, RefCell<Scaffold>>,
-    pub contentStringLen: c_uint,
+    paramEntities: HashMap<HashKey, Box<ENTITY>>,
+}
+
+pub struct DTDPools {
+    pool: STRING_POOL,
+    entityValuePool: STRING_POOL,
+}
+
+impl DTDPools {
+    fn new() -> DTDPools {
+        unsafe {
+            let mut pools = std::mem::zeroed::<DTDPools>();
+            pools.pool.init();
```

Comment:
Maybe worth noting that I got rid of the `init` and `destroy` methods from string pools in favor of `try_new` and `Drop`, respectively in #53.

Reply:
That's fine, I'll wait for you to merge that into `master`, then I'll merge that into this one.